fedimint-testing/src/gateway.rs

```diff
@@ -0,0 +1,70 @@
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use fedimint_client::module::gen::ClientModuleGenRegistry;
+use fedimint_core::module::registry::ModuleDecoderRegistry;
+use fedimint_core::task::TaskGroup;
+use ln_gateway::client::{DynGatewayClientBuilder, MemDbFactory, StandardGatewayClientBuilder};
+use ln_gateway::lnrpc_client::ILnRpcClient;
+use ln_gateway::rpc::rpc_client::RpcClient;
+use ln_gateway::rpc::ConnectFedPayload;
+use ln_gateway::Gateway;
+use tokio::sync::RwLock;
+use url::Url;
+
+use crate::federation::FederationTest;
+
+pub struct GatewayTest {
+    // TODO: make private and set the password in `RpcClient`
+    pub password: String,
+    address: Url,
+}
+
+impl GatewayTest {
+    /// RPC client for communicating with the gateway admin API
+    pub async fn new_client(&self) -> RpcClient {
+        RpcClient::new(self.address.clone())
+    }
+
+    pub async fn connect_fed(&self, fed: &FederationTest) {
+        let connect = fed.connection_code().to_string();
+        let client = self.new_client().await;
+        client
+            .connect_federation(self.password.clone(), ConnectFedPayload { connect })
+            .await
+            .unwrap();
+    }
+
+    pub(crate) async fn new(
+        base_port: u16,
+        lightning: impl ILnRpcClient + 'static,
+        decoders: ModuleDecoderRegistry,
+        module_gens: ClientModuleGenRegistry,
+    ) -> Self {
+        let listen: SocketAddr = format!("127.0.0.1:{base_port}").parse().unwrap();
+        let address: Url = format!("http://{listen}").parse().unwrap();
+        let mut task = TaskGroup::new();
+        let password = rand::random::<u64>().to_string();
+
+        // Create federation client builder for the gateway
+        let client_builder: DynGatewayClientBuilder =
+            StandardGatewayClientBuilder::new(PathBuf::new(), MemDbFactory.into(), address.clone())
+                .into();
+
+        let gateway = Gateway::new_with_lightning_connection(
+            Arc::new(RwLock::new(lightning)),
+            client_builder.clone(),
+            decoders.clone(),
+            module_gens.clone(),
+            task.clone(),
+        )
+        .await
+        .unwrap();
+
+        gateway.spawn_webserver(listen, password.clone()).await;
```

Comment:
Nit: Doesn't need to be this PR but this should probably be moved inside of `Gateway`.

Reply:
The fn is inside of `Gateway`.  We spawn it separately so we can know the API is up before continuing.  I do a similar thing for our federations.